Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Basic.lean

```diff
@@ -0,0 +1,427 @@
+ * $\Psi_n := \tilde{\Psi}_n\psi_2$ if $n$ is even, and
+ * $\Psi_n := \tilde{\Psi}_n$ if $n$ is odd.
+
+Furthermore, define the associated sequences $\Psi_n^{[2]}, \Phi_n \in R[X]$ by
+ * $\Psi_n^{[2]} := \tilde{\Psi}_n^2\Psi_2^{[2]}$ if $n$ is even,
+ * $\Psi_n^{[2]} := \tilde{\Psi}_n^2$ if $n$ is odd,
+ * $\Phi_n := X\Psi_n^{[2]} - \tilde{\Psi}_{n + 1}\tilde{\Psi}_{n - 1}$ if $n$ is even, and
+ * $\Phi_n := X\Psi_n^{[2]} - \tilde{\Psi}_{n + 1}\tilde{\Psi}_{n - 1}\Psi_2^{[2]}$ if $n$ is odd.
+
+With these definitions in mind, $\psi_n \in R[X, Y]$ and $\phi_n \in R[X, Y]$ are congruent in
+$R[W]$ to $\Psi_n \in R[X, Y]$ and $\Phi_n \in R[X]$ respectively, which are defined in terms of
+$\Psi_2^{[2]} \in R[X]$ and $\tilde{\Psi}_n \in R[X]$.
+
+## Main definitions
+
+ * `WeierstrassCurve.preΨ`: the univariate polynomials $\tilde{\Psi}_n$.
+ * `WeierstrassCurve.ΨSq`: the univariate polynomials $\Psi_n^{[2]}$.
+ * `WeierstrassCurve.Ψ`: the bivariate polynomials $\Psi_n$.
+ * `WeierstrassCurve.Φ`: the univariate polynomials $\Phi_n$.
+ * `WeierstrassCurve.ψ`: the bivariate $n$-division polynomials $\psi_n$.
+ * `WeierstrassCurve.φ`: the bivariate polynomials $\phi_n$.
+ * TODO: the bivariate polynomials $\omega_n$.
+
+## Implementation notes
+
+Analogously to `Mathlib.NumberTheory.EllipticDivisibilitySequence`, the $n$-division bivariate
+polynomials $\Psi_n$ are defined in terms of the univariate polynomials $\tilde{\Psi}_n$. This is
+done partially to avoid ring division, but more crucially to allow the definition of $\Psi_n^{[2]}$
+and $\Phi_n$ as univariate polynomials without needing to work under the coordinate ring, and to
+allow the computation of their leading terms without ambiguity. Furthermore, evaluating these
+polynomials at a rational point on $W$ recovers their original definition up to linear combinations
+of the Weierstrass equation of $W$, hence also avoiding the need to work under the coordinate ring.
+
+TODO: implementation notes for the definition of $\omega_n$.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, division polynomial, torsion point
+-/
+
+open Polynomial PolynomialPolynomial
+
+local macro "C_simp" : tactic =>
+  `(tactic| simp only [map_ofNat, C_0, C_1, C_neg, C_add, C_sub, C_mul, C_pow])
+
+universe u
+
+namespace WeierstrassCurve
+
+variable {R : Type u} [CommRing R] (W : WeierstrassCurve R)
+
+/-! ### The univariate polynomial $\Psi_2^{[2]}$ -/
+
+/-- The $2$-division polynomial $\psi_2 = \Psi_2$. -/
+protected noncomputable def ψ₂ : R[X][Y] :=
+  W.toAffine.polynomialY
+
+/-- The univariate polynomial $\Psi_2^{[2]}$ congruent to $\psi_2^2$. -/
+protected noncomputable def Ψ₂Sq : R[X] :=
+  C 4 * X ^ 3 + C W.b₂ * X ^ 2 + C (2 * W.b₄) * X + C W.b₆
+
+lemma C_Ψ₂Sq_eq : C W.Ψ₂Sq = W.ψ₂ ^ 2 - 4 * W.toAffine.polynomial := by
+  rw [WeierstrassCurve.Ψ₂Sq, WeierstrassCurve.ψ₂, b₂, b₄, b₆, Affine.polynomialY, Affine.polynomial]
```

Comment:
Even if you protect these, you can still `open WeierstrassCurve (Ψ₂Sq ψ₂)` etc. to get rid of the `WeierstrassCurve.`.